### PR TITLE
Avoid reporting CA1823 for inline arrays

### LINKED ISSUE #6789: Suppress CA1823 and related analyzers in [InlineArray] types
.NET 8 introduces the new InlineArrayAttribute, which is put onto a type to indicate that its sole field should be replicated the specified number of times.  C# then provides automatic helpers for indexing into these types and getting at their data as spans.  As a result, the field declared inside the type isn't actually referenced by the developer, and so various analyzers that warn about that end up firing false positives, e.g. CA1823 "Avoid unused private fields".  Such analyzers should skip types marked as InlineArrayAttribute.

### PR DESCRIPTION
**Pull request by sharwell**
Fixes #6789

### CODE REVIEW

File: src/Test.Utilities/AdditionalMetadataReferences.cs [hunk 1]
```diff
@@ -9,6 +10,13 @@ namespace Test.Utilities
 {
     public static class AdditionalMetadataReferences
     {
+        public static ReferenceAssemblies Net80 { get; } = new ReferenceAssemblies(
+            "net8.0",
+            new PackageIdentity(
+                "Microsoft.NETCore.App.Ref",
+                "8.0.0-preview.6.23329.7"),
```
Comment (stephentoub) [inline src/Test.Utilities/AdditionalMetadataReferences.cs:17]: Does this need to be updated whenever the repo updates the version it depends on preview to preview?

Comment (sharwell) [reply src/Test.Utilities/AdditionalMetadataReferences.cs:17]: Not really. Normally the signatures in the reference assemblies don't change to the point of affecting the unit tests so they just keep working. Plus, it'll end up in dotnet/roslyn-sdk soon.

Comment (mavasani) [reply src/Test.Utilities/AdditionalMetadataReferences.cs:17]: Another potential approach we take for tests is to add an additional source file to the test with custom stub implementation of the newer APIs added to the Framework which are not yet available via the currently referenced assemblies. We can do a sweeping cleanup later once we move the tests to reference packages with required APIs.

File: src/Utilities/Compiler/Extensions/ISymbolExtensions.cs [hunk 2]
```diff
@@ -659,6 +659,20 @@ public static bool HasAnyAttribute(this ISymbol symbol, params INamedTypeSymbol?
             return symbol.GetAttributes(attributeTypesToMatch).Any();
         }
 
+        public static bool HasAnyAttribute(this ISymbol symbol, [NotNullWhen(true)] INamedTypeSymbol? attributeToMatch)
```
Comment (mavasani) [inline src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:662]: What is the difference between this extension, and the one right below? https://github.com/dotnet/roslyn-analyzers/blob/1d9fd4e26f90c271b20560613808cac1745e7422/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L662-L680

**Review by mavasani — APPROVED** [review]
Changes LGTM, though we should consolidate all the `HasXXXAttribute` helpers instead of adding a new one.

File: src/Test.Utilities/AdditionalMetadataReferences.cs (hunk 1 above)
Comment (sharwell) [reply src/Test.Utilities/AdditionalMetadataReferences.cs:17]: > Another potential approach we take for tests is to add an additional source file to the test with custom stub implementation of the newer APIs

This is necessary when the APIs haven't been published, but in this case we can reference the real APIs and avoid the risk involved in cases where a deviation in our test implementation impacts the behavior of the analyzer.

File: src/Utilities/Compiler/Extensions/ISymbolExtensions.cs (hunk 2 above)
Comment (sharwell) [reply src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:662]: ➡️ I'll send a follow-up to address this.

Comment (sharwell) [reply src/Utilities/Compiler/Extensions/ISymbolExtensions.cs:662]: ➡️ To answer the question, the previous `HasAttribute` method was delegating to the `params` form of `HasAnyAttribute`. The new method avoids the allocation that all calls to `HasAttribute` were facing.

**Pull request merged**